File: leetcode/java/0125_valid-palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        StringBuffer b = new StringBuffer();
        for (int i = 0; i < s.length(); ++i) {
            char c = s.charAt(i);
            if (Character.isLetterOrDigit(c)) {
                char x = Character.toLowerCase(c);
                b.append(x);
            }
        }
        String w = b.toString();
        for (int i = 0, k = w.length() - 1; i < k; i++, k--) {
            if (w.charAt(i) != w.charAt(k)) {
                return false;
            }
        }
        return true;
    }
}

File: leetcode/cpp/0125_valid-palindrome.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        if (s.size() == 0) return true;
        string w;
        w.reserve(s.size());
        int i;
        for (i=0; i<s.size(); ++i) {
            if (s[i] >= 'A' && s[i] <= 'Z') {
                w.push_back(s[i] - 'A' + 'a');
            } else if (s[i] >= 'a' && s[i] <= 'z') {
                w.push_back(s[i]);
            } else if (s[i] >= '0' && s[i] <= '9') {
                w.push_back(s[i]);
            }
        }
        return string(w.rbegin(), w.rend()) == w;
    }
};